File: 0743-network-delay-time/0743-network-delay-time.cpp
class Solution {
public:
    int networkDelayTime(vector<vector<int>>& times, int n, int k) {
        vector<int>dist(n,INT_MAX);
        dist[k-1]=0;
        for(int i=0;i<n-1;i++){

            for(auto ed:times){
                int u=ed[0]-1,v=ed[1]-1,t=ed[2];

                if(dist[u]!=INT_MAX && dist[u]+t<dist[v]){
                    dist[v]=dist[u]+t;
                }
            }
        }
        int total=0;
        for(int i=0;i<n;i++){
            if(dist[i]==INT_MAX) return -1;
            total=max(total,dist[i]);
        }
        return total;
        
    }
};

File: 0743-network-delay-time/0743-network-delay-time.py
import heapq

class Solution:
    def networkDelayTime(self, times: list[list[int]], n: int, k: int) -> int:

        adj = [[] for _ in range(n + 1)]

        for u, v, w in times:
            adj[u].append((v, w))

        dist = [float('inf')] * (n + 1)
        dist[k] = 0

        pq = [(0, k)]

        while pq:

            curr_dist, node = heapq.heappop(pq)

            if curr_dist > dist[node]:
                continue

            for neighbour, weight in adj[node]:

                if curr_dist + weight < dist[neighbour]:

                    dist[neighbour] = curr_dist + weight

                    heapq.heappush(
                        pq,
                        (dist[neighbour], neighbour)
                    )

        ans = 0

        for i in range(1, n + 1):

            if dist[i] == float('inf'):
                return -1

            ans = max(ans, dist[i])

        return ans